vcd/resource_vcd_vdc.go

```diff
@@ -0,0 +1,423 @@
+			},
+		},
+	}
+}
+
+func resourceVcdVdcCreate(d *schema.ResourceData, meta interface{}) error {
+	log.Printf("[TRACE] vdc creation initiated")
+
+	vcdClient := meta.(*VCDClient)
+
+	// vdc creation is accessible only in administrator API part
+	adminOrg, err := vcdClient.GetAdminOrgFromResource(d)
+	if err != nil {
+		return fmt.Errorf(errorRetrievingOrg, err)
+	}
+
+	params, err := getVcdVdcInput(d, vcdClient)
+	if err != nil {
+		return err
+	}
+
+	task, err := adminOrg.CreateVdc(params)
+	if err != nil {
+		log.Printf("[DEBUG] Error creating vdc: %#v", err)
+		return fmt.Errorf("error creating vdc: %#v", err)
+	}
+
+	d.SetId(d.Get("name").(string))
+	log.Printf("[TRACE] vdc created: %#v", task)
+	return resourceVcdVdcRead(d, meta)
+}
+
+func resourceVcdVdcRead(d *schema.ResourceData, meta interface{}) error {
+	log.Printf("[TRACE] vdc read initiated")
+
+	vcdClient := meta.(*VCDClient)
+
+	adminOrg, err := vcdClient.GetAdminOrgFromResource(d)
+	if err != nil {
+		return fmt.Errorf(errorRetrievingOrg, err)
+	}
+
+	vdc, err := adminOrg.GetVdcByName(d.Id())
+	if err != nil || vdc == (govcd.Vdc{}) {
+		log.Printf("[DEBUG] Unable to find vdc. Removing from tfstate")
+		d.SetId("")
+		return nil
+	}
+
+	log.Printf("[TRACE] vdc read completed: %#v", vdc.Vdc)
+	return nil
+}
+
+//update function for "delete_force", "delete_recursive" no actions needed
+func resourceVcdVdcUpdate(d *schema.ResourceData, m interface{}) error {
+	return nil
+}
+
+func resourceVcdVdcDelete(d *schema.ResourceData, meta interface{}) error {
+	log.Printf("[TRACE] vdc delete started")
+
+	vcdClient := meta.(*VCDClient)
+
+	adminOrg, err := vcdClient.GetAdminOrgFromResource(d)
+	if err != nil {
+		return fmt.Errorf(errorRetrievingOrg, err)
+	}
+
+	vdc, err := adminOrg.GetVdcByName(d.Id())
+	if err != nil || vdc == (govcd.Vdc{}) {
+		log.Printf("[DEBUG] Unable to find vdc. Removing from tfstate")
+		d.SetId("")
+		return nil
+	}
+
+	_, err = vdc.Delete(d.Get("delete_force").(bool), d.Get("delete_recursive").(bool))
+	if err != nil {
+		log.Printf("[DEBUG] Error removing vdc %#v", err)
+		return fmt.Errorf("error removing vdc %#v", err)
+	}
+
+	log.Printf("[TRACE] vdc delete completed: %#v", vdc.Vdc)
+	return nil
+}
+
+func capacityWithUsage(d map[string]interface{}) *types.CapacityWithUsage {
+	v := &types.CapacityWithUsage{
+		Units: d["units"].(string),
+	}
+
+	if allocated, ok := d["allocated"]; ok {
+		v.Allocated = int64(allocated.(int))
+	}
+
+	if limit, ok := d["limit"]; ok {
+		v.Limit = int64(limit.(int))
+	}
+
+	if reserved, ok := d["reserved"]; ok {
+		v.Reserved = int64(reserved.(int))
+	}
+
+	if used, ok := d["used"]; ok {
+		v.Used = int64(used.(int))
+	}
+
+	if overhead, ok := d["overhead"]; ok {
+		v.Overhead = int64(overhead.(int))
+	}
+
+	return v
+}
+
+func getVcdVdcInput(d *schema.ResourceData, vcdClient *VCDClient) (*types.VdcConfiguration, error) {
```

Comment:
Some comments are needed here.
What are we extracting from the resource data and what do we want to achieve?